RESTEasy Reactive - Support text/binary conversion for multipart files

Sgitario opened the pull request:
Assuming we receive the following request:
```
Content-Disposition: form-data; name="textPart"; filename="my_file.txt"
Content-Type: application/octet-stream
Content-Transfer-Encoding: binary
```

We can load the content of the file using the `FileUpload` class:

```java
public class FormData {
    @FormParam("myFile")
    public FileUpload myFile;
}
```

However, according to what the specs state:
```
The presence of the filename parameter does not force an
implementation to write the entity to a separate file. It is
perfectly acceptable for implementations to leave the entity as part
of the normal mail stream unless the user requests otherwise. As a
consequence, the parameter may be used on any MIME entity, even
`inline' ones. These will not normally be written to files, but the
parameter could be used to provide a filename if the receiving user
should choose to write the part to a file.
```

Then, we should not enforce the use of `FileUpload` and hence also support reading the whole content of the file into strings, and other binary formats like byte[] and InputStream.

```java
public class FormData {
    @FormParam("myFile")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public String myFile;
}

public class FormData {
    @FormParam("myFile")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public byte[] binaryPart;
}

public class FormData {
    @FormParam("myFile")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public InputStream streamPart;
}
```

Fix https://github.com/quarkusio/quarkus/issues/27083

geoand commented on independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java:
```diff
@@ -867,12 +868,20 @@ public Object getFormParameter(String name, boolean single, boolean encoded) {
             }
-            if (encoded) {
+
+            if (val.isFileItem()) {
```
What case does this change cater to?

Sgitario replied:
When using:
```java
public class FormData {
    @FormParam("myFile")
    public String myFile;
}

```

Sgitario replied:
One idea is to force users to use `@PartType(MediaType.APPLICATION_OCTET_STREAM)` to copy the file content to a string. 
However, as far I understood, this is how resteasy classic works and what the spec says :/

geoand replied:
The spec doesn't say anything about `text` files though, does it? 
All I am questioning here is the need to support `String`. I would much rather not support it initially and have users manually convert the string if necessary.
Or we really want `String`, I think it only makes sense to do so for certain media types.

Sgitario replied:
I've just confirmed that Resteasy Classic works like this. 
If users try to load a binary file to String, I would say this is a user issue, but the framework should not stop this use case. 
Having said that and despite what Resteasy Classic does, I agree with you that we should force users to use the media type `@PartType(MediaType.APPLICATION_OCTET_STREAM)` if they want to copy the content of a file to a String. 

Let me know if you want me to do this.

geoand replied:
Sounds good

Sgitario replied:
PR updated

Sgitario committed:
RESTEasy Reactive - Support text/binary conversion for multipart files

Assuming we receive the following request:
```
Content-Disposition: form-data; name="textPart"; filename="my_file.txt"
Content-Type: application/octet-stream
Content-Transfer-Encoding: binary
```

We can load the content of the file using the `FileUpload` class:

```java
public class FormData {
    @FormParam("myFile")
    public FileUpload myFile;
}
```

However, according to what the specs state:
```
The presence of the filename parameter does not force an
implementation to write the entity to a separate file. It is
perfectly acceptable for implementations to leave the entity as part
of the normal mail stream unless the user requests otherwise. As a
consequence, the parameter may be used on any MIME entity, even
`inline' ones. These will not normally be written to files, but the
parameter could be used to provide a filename if the receiving user
should choose to write the part to a file.
```

Then, we should not enforce the use of `FileUpload` and hence also support reading the whole content of the file into strings, and other binary formats like byte[] and InputStream.

```java
public class FormData {
    @FormParam("myFile")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public String myFile;
}

public class FormData {
    @FormParam("myFile")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public byte[] binaryPart;
}

public class FormData {
    @FormParam("myFile")
    @PartType(MediaType.APPLICATION_OCTET_STREAM)
    public InputStream streamPart;
}
```

Fix https://github.com/quarkusio/quarkus/issues/27083
